Source/FieldSolver/SpectralSolver/SpectralAlgorithms/PsatdAlgorithm.cpp

```diff
@@ -226,9 +250,83 @@ PsatdAlgorithm::pushSpectralFields (SpectralFieldData& f) const
                                     - I * T2 * S_ck * (kx * Ey_old - ky * Ex_old)
                                     + I * X1 * (kx * Jy - ky * Jx);
 
-            // Additional update equations for averaged Galilean algorithm
+            if (J_linear_in_time)
+            {
+                const Complex Jx_new = fields(i,j,k,IdxLin::Jx_new);
+                const Complex Jy_new = fields(i,j,k,IdxLin::Jy_new);
+                const Complex Jz_new = fields(i,j,k,IdxLin::Jz_new);
+
+                const Complex F_old = fields(i,j,k,IdxLin::F);
+                const Complex G_old = fields(i,j,k,IdxLin::G);
+
+                fields(i,j,k,Idx::Ex) += -X1 * (Jx_new - Jx) / dt + I * c2 * S_ck * F_old * kx;
+
+                fields(i,j,k,Idx::Ey) += -X1 * (Jy_new - Jy) / dt + I * c2 * S_ck * F_old * ky;
+
+                fields(i,j,k,Idx::Ez) += -X1 * (Jz_new - Jz) / dt + I * c2 * S_ck * F_old * kz;
+
+                fields(i,j,k,Idx::Bx) += I * X2/c2 * (ky * (Jz_new - Jz) - kz * (Jy_new - Jy));
+                    + I * c2 * S_ck * G_old * kx;
+
+                fields(i,j,k,Idx::By) += I * X2/c2 * (kz * (Jx_new - Jx) - kx * (Jz_new - Jz));
+                    + I * c2 * S_ck * G_old * ky;
+
+                fields(i,j,k,Idx::Bz) += I * X2/c2 * (kx * (Jy_new - Jy) - ky * (Jx_new - Jx));
+                    + I * c2 * S_ck * G_old * kz;
+
+                const Complex k_dot_J  = kx * Jx + ky * Jy + kz * Jz;
+                const Complex k_dot_dJ = kx * (Jx_new - Jx) + ky * (Jy_new - Jy) + kz * (Jz_new - Jz);
+                const Complex k_dot_E = kx * Ex_old + ky * Ey_old + kz * Ez_old;
+                const Complex k_dot_B = kx * Bx_old + ky * By_old + kz * Bz_old;
 
-            if (time_averaging)
+                fields(i,j,k,IdxLin::F) = C * F_old + S_ck * (I * k_dot_E - rho_old * inv_ep0)
+                    - X1 * ((rho_new - rho_old) / dt + I * k_dot_J) - I * X2/c2 * k_dot_dJ;
+
+                fields(i,j,k,IdxLin::G) = C * G_old + I * S_ck * k_dot_B;
+
+                if (time_averaging)
+                {
+                    const Complex X5 = X5_arr(i,j,k);
+                    const Complex X6 = X6_arr(i,j,k);
+
+                    // TODO: Here the code is *accumulating* the average,
+                    // because it is meant to be used with sub-cycling
+                    // maybe this should be made more generic
+
+                    fields(i,j,k,IdxLin::Ex_avg) += S_ck * Ex_old
+                        + I * c2 * ep0 * X1 * (ky * Bz_old - kz * By_old)
+                        + I * X5 * rho_old * kx + I * X6 * rho_new * kx
+                        + X3/c2 * Jx - X2/c2 * Jx_new + I * c2 * ep0 * X1 * F_old * kx;
```

Comment:
I can certainly do that, however I have a question before I proceed:
What is the "rule of thumb" about these kinds of temporary variables in device kernels?

I thought that using too many temporary variables in device code might degrade the overall performance of the kernel. I'm asking because there are several other combinations that could be stored in temporary variables in principle, but I thought it would be faster to re-compute those on the fly each time.

Reply:
That's a good question and best answered by looking at the registers. Compile with `-DAMReX_CUDA_PTX_VERBOSE=ON` and change only the lines in the kernel to compare.

Since we multipl 3-4 values here that also themselves come from registers and use them ~3 times each I would expect that this can save some registers by accessing them once and placing them as intermediate result. This might be counter-acted if the individual values are then again used individually...

So the rule of thumb is: if you intermediate result is constants like here, it's not obvious. But if you intermediate result contains an access to global memory, e.g. term `A * field(i,j,k)`, then the situation is almost always better to store instead of re-compute.